### DISCUSSION TITLE
SQS Reject Task on Sigterm?

### DISCUSSION THREAD

**Discussion by Asafb26** [celery/celery#8083] [Q&A]
On an ECS fargate server, we have some long, heavy tasks triggered by an SQS queue.
We have set our `visibility_timeout` to 12 hours, and enabled `task_acks_late` and `task_reject_on_worker_lost`.
As we deploy a new revision of this service, we have limited time for graceful shutdown (2 minutes is the maximum time window for a fargate task), so we need to wait 12 hours before requeuing long tasks.
There is an option to reject a task in Celery, but is there a setting to reject all tasks (from all processes) on SIGTERM?

**Comment by pbudzon** [comment]
Celery tries to perform a "warm shutdown" on SIGTERM where it waits for all tasks for finish. Obviously this doesn't work when you have a time limit, like in your case. You can ask celery to do a "cold shutdown" by sending it a SIGQUIT signal instead - that will cause it to abort all running jobs and return the tasks to the queue. ECS doesn't have an option to change which signal is sent on shutdown (like you can do with `docker container kill --signal xxx` or with `STOPSIGNAL` in Dockerfile), but apparently there's a hidden "feature" in celery implemented originally for Heroku (for this exact scenario, actually).

**Set environment variable `REMAP_SIGTERM=SIGQUIT` in your ECS (or any docker) container and that will cause celery to perform a cold shutdown (return tasks to queue immediately) on SIGTERM instead of a warm shutdown (wait for tasks to complete).**

It's slightly annoying that this is not documented - you can find it if you comb through PRs and Issues on GitHub, but it's not in the docs. Even though this seems like a very useful feature for anyone running celery in docker.

**Comment by Nusnus** [comment]
@pbudzon 
> ... but apparently there's a hidden "feature" in celery implemented originally for Heroku (for this exact scenario, actually).
> 
> It's slightly annoying that this is not documented - you can find it if you comb through PRs and Issues on GitHub, but it's not in the docs. Even though this seems like a very useful feature for anyone running celery in docker.

Thank you for this comment. It will no longer be hidden from now on :)

![CleanShot 2024-08-30 at 16 35 21@2x](https://github.com/user-attachments/assets/fa49a4fe-7630-4ff8-908d-8aa743ecbc14)

https://github.com/celery/celery/pull/9213

@Asafb26 FYI
